File: packages/std/native/hash.c
// EzLang std/hash 原生封装层

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    uint8_t *data;
    int64_t size;
} Blob;

static const uint8_t *ez_blob_data(const Blob *data, size_t *size) {
    if (!data || data->size <= 0 || !data->data || (uint64_t)data->size > (uint64_t)SIZE_MAX) {
        *size = 0;
        return (const uint8_t *)"";
    }
    *size = (size_t)data->size;
    return data->data;
}

uint32_t hashFnv1a32(const Blob *data) {
    size_t size = 0;
    const uint8_t *bytes = ez_blob_data(data, &size);
    uint32_t hash = 2166136261u;
    for (size_t i = 0; i < size; ++i) {
        hash ^= bytes[i];
        hash *= 16777619u;
    }
    return hash;
}

uint64_t hashFnv1a64(const Blob *data) {
    size_t size = 0;
    const uint8_t *bytes = ez_blob_data(data, &size);
    uint64_t hash = 14695981039346656037ULL;
    for (size_t i = 0; i < size; ++i) {
        hash ^= bytes[i];
        hash *= 1099511628211ULL;
    }
    return hash;
}

uint32_t hashStrFnv1a32(const char *s) {
    Blob data = {(uint8_t *)(s ? s : ""), (int64_t)strlen(s ? s : "")};
    return hashFnv1a32(&data);
}

uint64_t hashStrFnv1a64(const char *s) {
    Blob data = {(uint8_t *)(s ? s : ""), (int64_t)strlen(s ? s : "")};
    return hashFnv1a64(&data);
}

uint64_t hashCombineU64(uint64_t seed, uint64_t value) {
    return seed ^ (value + 0x9E3779B97F4A7C15ULL + (seed << 6) + (seed >> 2));
}

uint32_t crc32(const Blob *data) {
    size_t size = 0;
    const uint8_t *bytes = ez_blob_data(data, &size);
    uint32_t crc = 0xFFFFFFFFu;
    for (size_t i = 0; i < size; ++i) {
        crc ^= bytes[i];
        for (int bit = 0; bit < 8; ++bit) {
            uint32_t mask = (uint32_t)(-(int32_t)(crc & 1u));
            crc = (crc >> 1) ^ (0xEDB88320u & mask);
        }
    }
    return ~crc;
}

uint32_t crc32Str(const char *s) {
    Blob data = {(uint8_t *)(s ? s : ""), (int64_t)strlen(s ? s : "")};
    return crc32(&data);
}

File: packages/std/emcc/hash.js
// EzLang std/hash Emscripten JS 封装层
(function () {
  function blobBytes(blobPtr) {
    if (!blobPtr) return new Uint8Array(0);
    var dataPtr = getValue(blobPtr, '*');
    var size = Number(getValue(blobPtr + 8, 'i64'));
    if (!Number.isFinite(size) || size <= 0 || Math.floor(size) !== size) return new Uint8Array(0);
    if (!dataPtr || dataPtr < 0 || dataPtr > HEAPU8.length || size > HEAPU8.length - dataPtr) return new Uint8Array(0);
    return HEAPU8.slice(dataPtr, dataPtr + size);
  }

  function cStringBytes(ptr) {
    if (!ptr) return new Uint8Array(0);
    var end = ptr;
    while (HEAPU8[end] !== 0) end++;
    return HEAPU8.slice(ptr, end);
  }

  function fnv1a32(bytes) {
    var hash = 2166136261 >>> 0;
    for (var i = 0; i < bytes.length; i++) {
      hash ^= bytes[i];
      hash = Math.imul(hash, 16777619) >>> 0;
    }
    return hash | 0;
  }

  function fnv1a64(bytes) {
    var hash = 14695981039346656037n;
    for (var i = 0; i < bytes.length; i++) {
      hash ^= BigInt(bytes[i]);
      hash = BigInt.asUintN(64, hash * 1099511628211n);
    }
    return BigInt.asIntN(64, hash);
  }

  function crc32Bytes(bytes) {
    var crc = 0xFFFFFFFF >>> 0;
    for (var i = 0; i < bytes.length; i++) {
      crc = (crc ^ bytes[i]) >>> 0;
      for (var bit = 0; bit < 8; bit++) {
        var mask = -(crc & 1);
        crc = ((crc >>> 1) ^ (0xEDB88320 & mask)) >>> 0;
      }
    }
    return (~crc) | 0;
  }

  mergeInto(LibraryManager.library, {
    hashFnv1a32: function (dataPtr) {
      return fnv1a32(blobBytes(dataPtr));
    },
    hashFnv1a64: function (dataPtr) {
      return fnv1a64(blobBytes(dataPtr));
    },
    hashStrFnv1a32: function (s) {
      return fnv1a32(cStringBytes(s));
    },
    hashStrFnv1a64: function (s) {
      return fnv1a64(cStringBytes(s));
    },
    hashCombineU64: function (seed, value) {
      seed = BigInt.asUintN(64, BigInt(seed));
      value = BigInt.asUintN(64, BigInt(value));
      var out = seed ^ (value + 0x9E3779B97F4A7C15n + (seed << 6n) + (seed >> 2n));
      return BigInt.asIntN(64, BigInt.asUintN(64, out));
    },
    crc32: function (dataPtr) {
      return crc32Bytes(blobBytes(dataPtr));
    },
    crc32Str: function (s) {
      return crc32Bytes(cStringBytes(s));
    },
  });
})();
